oss-internship-2020/libtiff/test/defer_strile_writing.cc

```diff
@@ -0,0 +1,318 @@
+          << "TIFFWriteEncodedStrip fatal error";
+      EXPECT_THAT(status_or_int.value(), Eq(1))
+          << "line " << i << ": expected 1, got " << status_or_int.value();
+
+      if (i == 1 && height > 100000) {
+        i = height - 2;
+      }
+    }
+  }
+
+  ASSERT_THAT(api.TIFFClose(&tif), IsOk()) << "TIFFClose fatal error";
+
+  sapi::v::ConstCStr r_var("r");
+  status_or_tif = api.TIFFOpen(srcfile_var.PtrBefore(), r_var.PtrBefore());
+  ASSERT_THAT(status_or_tif, IsOk()) << "Could not open " << srcfile;
+
+  sapi::v::RemotePtr tif2(status_or_tif.value());
+  ASSERT_THAT(tif2.GetValue(), NotNull()) << "can't open " << srcfile;
+
+  if (tiled) {
+    for (int i = 0; i < (height + 15) / 16; ++i) {
+      for (int retry = 0; retry < 2; ++retry) {
+        std::array<uint8_t, kTileBufferSize> tilebuffer;
+        uint8_t expected_c = static_cast<uint8_t>(i);
+        tilebuffer.fill(0);
+
+        sapi::v::Array<uint8_t> tilebuffer_(tilebuffer.data(),
+                                            tilebuffer.size());
+
+        status_or_long = api.TIFFReadEncodedTile(
+            &tif2, i, tilebuffer_.PtrBoth(), kTileBufferSize);
+        ASSERT_THAT(status_or_long, IsOk())
+            << "TIFFReadEncodedTile fatal error";
+        EXPECT_THAT(status_or_long.value(), Eq(kTileBufferSize))
+            << "line " << i << ": expected " << kTileBufferSize << ", got "
+            << status_or_long.value();
+
+        bool cmp = tilebuffer[0] != expected_c || tilebuffer[255] != expected_c;
+
+        EXPECT_THAT(tilebuffer[0], Eq(expected_c))
+            << "unexpected value at tile " << i << ": expected " << expected_c
+            << ", got " << tilebuffer[0];
+
+        EXPECT_THAT(tilebuffer[255], Eq(expected_c))
+            << "unexpected value at tile " << i << ": expected " << expected_c
+            << ", got " << tilebuffer[255];
+        if (cmp) {
+          ASSERT_THAT(api.TIFFClose(&tif2), IsOk()) << "TIFFClose fatal error";
+        }
+      }
+    }
+  } else {
+    for (int i = 0; i < height; ++i) {
+      for (int retry = 0; retry < 2; ++retry) {
+        sapi::v::UChar c(0);
+        uint8_t expected_c = static_cast<uint8_t>(i);
+
+        status_or_long = api.TIFFReadEncodedStrip(&tif2, i, c.PtrBoth(), 1);
+        ASSERT_THAT(status_or_long, IsOk())
+            << "TIFFReadEncodedStrip fatal error";
+        EXPECT_THAT(status_or_long.value(), Eq(1))
+            << "line " << i << ": expected 1, got " << status_or_long.value();
+        EXPECT_THAT(c.GetValue(), Eq(expected_c))
+            << "unexpected value at line " << i << ": expected " << expected_c
+            << ", got " << c.GetValue();
+
+        if (c.GetValue() != expected_c) {
+          ASSERT_THAT(api.TIFFClose(&tif2), IsOk()) << "TIFFClose fatal error";
+        }
+      }
+    }
+  }
+
+  ASSERT_THAT(api.TIFFClose(&tif2), IsOk()) << "TIFFClose fatal error";
+
+  unlink(srcfile.c_str());
+}
+
+TEST(SandboxTest, DeferStrileWriting) {
```

Comment:
Seems like a good use-case for a "value-parameterized test":
https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests